web3c/oasis.js

```diff
@@ -1,26 +1,144 @@
+   */
+  _setupRpcs(web3) {
+    let methods = [
+      // Second parameter - the long-term key - is intercepted by the provider.
+      new web3.extend.Method({
+        name: 'getPublicKey',
+        call: 'confidential_getPublicKey',
+        params: 1,
+        inputFormatter: [web3.extend.formatters.inputAddressFormatter],
+        outputFormatter: getPublicKeyOutputFormatter
+      }),
+      new web3.extend.Method({
+        name: 'call',
+        call: 'confidential_call_enc',
+        params: 2,
+        inputFormatter: [
+          web3.extend.formatters.inputCallFormatter,
+          web3.extend.formatters.inputDefaultBlockNumberFormatter
+        ],
+        outputFormatter: callOutputFormatter
+      }),
+      new web3.extend.Method({
+        name: 'expiry',
+        call: 'oasis_getExpiry',
+        params: 1,
+        inputFormatter: [(address) => {
+          if (!address) {
+            throw Error('An address must be provided to retrieve contract expiry');
+          }
+          return web3.extend.formatters.inputAddressFormatter(address);
+        }],
+        outputFormatter: (res) => res
+      })
+    ];
+
+    methods.forEach((method) => {
+      method.setRequestManager(web3._requestManager);
+      method.attachToObject(this);
+    });
   }
 
-  _setupExpiry(web3) {
-    let expiry = new web3.extend.Method({
-      name: 'expiry',
-      call: 'oasis_getExpiry',
-      params: 1,
-      inputFormatter: [(address) => {
-        if (!address) {
-          throw Error('An address must be provided to retrieve contract expiry');
-        }
-        return web3.extend.formatters.inputAddressFormatter(address);
-      }],
-      outputFormatter: (res) => res
+  /**
+   * Creates the Oasis `Contract` constructor on the `oasis` namespace.
+   * The `Contract` behaves in the same way as the eth `Contract` object, except
+   * that it adds the ability to speciy custom properties likes expriy and
+   * confidentiality. If confidentiality is set, then transparently encrypts/decrypts
+   * all outgoing/incoming requests according to the web3c spec.
+   */
+  _setupContract(web3, storage, mraebox) {
+    // Save `this` so that we can refer to it and its properties inside `ConfidentialContract`.
+    // Otherwise `this` is overridden when `new` is used in `new Contract`.
+    let self = this;
+
+    this.Contract = makeContractFactory(web3, (address, options) => {
+      let provider = new OasisProvider(this.keyManager, web3._requestManager);
+
+      let keymanager = self.keyManager;
+
+      if (options && options.saveSession === false) {
+        keymanager = new KeyManager(web3, undefined, mraebox);
+        provider = new OasisProvider(keymanager, web3._requestManager);
+      }
+
+      if (options && options.key) {
+        keymanager.add(address, options.key);
+      }
+
+      if (address) {
+        keymanager.get(address, (public_key) => {
+          // TODO: update once peter merges in PR to turn error into an option.
```

Comment:
As the comment mentions, the getPublicKey endpoint currently errors. This is why we check `public_key === 'object'`. @peterjgilbert has a change going in soon to return a None type instead. So this will change to check that, instead.